src/bgw/scheduler.c

```diff
@@ -325,19 +330,124 @@ scheduled_bgw_job_start(ScheduledBgwJob *sjob, register_background_worker_callba
 	}
 }
 
-static List *
-initialize_scheduled_jobs_list(MemoryContext mctx)
+static void
+terminate_and_cleanup_job(ScheduledBgwJob *sjob)
 {
-	List	   *jobs = bgw_job_get_all(sizeof(ScheduledBgwJob), mctx);
-	ListCell   *lc;
+	if (sjob->handle != NULL)
+	{
+		TerminateBackgroundWorker(sjob->handle);
+		WaitForBackgroundWorkerShutdown(sjob->handle);
+		pfree(sjob->handle);
+	}
+	sjob->may_need_mark_end = false;
+	worker_state_cleanup(sjob);
+}
 
-	foreach(lc, jobs)
+/*
+ *  Update the given job list with whatever is in the bgw_job table. For overlapping jobs,
+ *  copy over any existing scheduler info from the given jobs list.
+ *  Assume that both lists are ordered by job ID.
+ *  Note that this function call will destroy cur_jobs_list and return a new list.
+ */
+List *
+update_scheduled_jobs_list(List *cur_jobs_list, MemoryContext mctx)
+{
+	List	   *new_jobs = bgw_job_get_all(sizeof(ScheduledBgwJob), mctx);
+	ListCell   *new_ptr = list_head(new_jobs);
+	ListCell   *cur_ptr = list_head(cur_jobs_list);
+
+	while (cur_ptr != NULL)
 	{
-		ScheduledBgwJob *sjob = lfirst(lc);
+		ScheduledBgwJob *new_sjob;
+		ScheduledBgwJob *cur_sjob;
+
+		if (new_ptr == NULL)
+			break;
+
+		new_sjob = lfirst(new_ptr);
+		cur_sjob = lfirst(cur_ptr);
+
+		if (cur_sjob->job.fd.id < new_sjob->job.fd.id)
+		{
+			/*
+			 * We don't need cur_sjob anymore. Make sure to clean up the job
+			 * state. Then keep advancing cur pointer until we catch up.
+			 */
+			while (cur_sjob->job.fd.id < new_sjob->job.fd.id)
+			{
+				/* Terminate and cleanup the job's BGW */
+				terminate_and_cleanup_job(cur_sjob);
+
+				cur_ptr = lnext(cur_ptr);
+				if (cur_ptr == NULL)
+					goto finish_new_list;
+				cur_sjob = lfirst(cur_ptr);
+			}
+		}
+		if (cur_sjob->job.fd.id == new_sjob->job.fd.id)
+		{
+			/*
+			 * Then this job already exists. Copy over any state and advance
+			 * both pointers.
+			 */
+			cur_sjob->job = new_sjob->job;
+			*new_sjob = *cur_sjob;
+
+			cur_ptr = lnext(cur_ptr);
+			new_ptr = lnext(new_ptr);
+		}
+		else if (cur_sjob->job.fd.id > new_sjob->job.fd.id)
+		{
+			/* Advance the new_job list until we catch up to cur_list */
```

Comment:
don't we need to scheduled_bgw_job_transition_state_to(lfirst(new_ptr), JOB_STATE_SCHEDULED); here?

Reply:
good point, done. The tests don't actually catch this because, in practice, because job IDs are monotonically increasing, this should never happen